src/Simulation/EntryPointDriver/Driver.cs

```diff
@@ -241,146 +316,241 @@ private void AddOptionIfAvailable<T>(Command command, OptionInfo<T> option)
                 }
 
                 return default;
-            });
-    }
-
-    /// <summary>
-    /// Static members for <see cref="Driver{TCallable,TIn,TOut}"/>.
-    /// </summary>
-    internal static class Driver
-    {
-        // TODO: Define the aliases as constants.
-
-        /// <summary>
-        /// The subscription option.
-        /// </summary>
-        internal static readonly OptionInfo<string> SubscriptionOption = new OptionInfo<string>(
-            ImmutableList.Create("--subscription"), "The subscription ID.");
+            };
 
-        /// <summary>
-        /// The resource group option.
-        /// </summary>
-        internal static readonly OptionInfo<string> ResourceGroupOption = new OptionInfo<string>(
-            ImmutableList.Create("--resource-group"), "The resource group name.");
+            command.AddValidator(validator);
+            return new[] { validator };
+        }
 
         /// <summary>
-        /// The workspace option.
+        /// Creates the simulate command.
         /// </summary>
-        internal static readonly OptionInfo<string> WorkspaceOption = new OptionInfo<string>(
-            ImmutableList.Create("--workspace"), "The workspace name.");
+        /// <param name="entryPointCommands">The entry point commands that will be the sub commands to the created command.</param>
+        /// <returns>The created simulate command with the validators for that command.</returns>
+        private static CommandWithValidators CreateSimulateCommand(
+            IEnumerable<CommandWithValidators> entryPointCommands)
+        {
+            var simulate = new Command("simulate", "(default) Run the program using a local simulator.");
+            if (entryPointCommands.Count() == 1)
+            {
+                var (epCommand, validators) = entryPointCommands.First();
```

Comment:
`CreateSimulateCommand` and `CreateSubmitCommand` should probably also take in `IReadOnlyCollection<CommandWithValidators>`.

Reply:
Is the idea to just replace all the instances of `IEnumerable` with `IReadOnlyCollection`?

Reply:
`IEnumerable` is good if you're enumerating a sequence of items once from start to finish, but enumerating it more than once can cause the sequence to be recomputed (e.g. if it came from a LINQ expression). See https://www.jetbrains.com/help/resharper/PossibleMultipleEnumeration.html.

Also, in general, `IEnumerable.Count()` has to enumerate the whole collection to count it (O(n)) - it has a special check to avoid that for `IEnumerable`s that happen to be `ICollection`s, but since `IReadOnlyCollection` comes with its own `Count` property, I think it's clearer to use that.

Reply:
I've replace all the references of IEnumerable (except one where it is the type provided by the System.CommandLine) with List types. This addresses the issue of the repeated computations, and List also has its own dedicated Count implementation that is constant time.